Mathlib/Tactic/ProdAssoc.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2023 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Adam Topaz
+-/
+import Mathlib.Lean.Expr.Basic
+import Mathlib.Logic.Equiv.Basic
+
+/-!
+# Associativity of products
+
+This file constructs a term elaborator for "obvious" equivalences between iterated products.
+For example,
+```lean
+(prod_assoc% : (α × β) × (γ × δ) ≃ α × (β × γ) × δ)
+```
+gives the "obvious" equivalence between `(α × β) × (γ × δ)` and `α × (β × γ) × δ`.
+-/
+
+namespace Lean.Expr
+
+open Lean Meta
+
+/-- A helper type to keep track of universe levels and types in iterated produts. -/
+inductive ProdTree where
+  | type (tp : Expr) (l : Level)
+  | prod (fst snd : ProdTree) (lfst lsnd : Level)
+deriving Repr
+
+/-- The iterated product corresponding to a `ProdTree`. -/
+def ProdTree.getType : ProdTree → Expr
+  | type tp _ => tp
+  | prod fst snd u v => mkAppN (.const ``Prod [u,v]) #[fst.getType, snd.getType]
+
+/-- The number of types appearing in an iterated product encoded as a `ProdTree`. -/
+def ProdTree.size : ProdTree → Nat
+  | type _ _ => 1
+  | prod fst snd _ _ => fst.size + snd.size
+
+/-- The components of an interated product, presented as a `ProdTree`. -/
+def ProdTree.components : ProdTree → List Expr
+  | type tp _ => [tp]
+  | prod fst snd _ _ => fst.components ++ snd.components
+
+/-- Make a `ProdTree` out of an `Expr`. -/
+def mkProdTree (e : Expr) : MetaM ProdTree :=
+  match e with
+    | .app (.app (.const ``Prod [u,v]) X) Y => do
+        return .prod (← X.mkProdTree) (← Y.mkProdTree) u v
+    | X => do
+      let some u := (← inferType X).type? | throwError "Not a type{indentExpr X}"
```

Comment:
I think you should do `(← whnfD (← inferType X)).type?` here since the type might not literally be `Type _`, and it might need a bit of reduction.

Reply:
Done. Thanks!